rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -117,13 +241,222 @@ class IfExprTree extends PostOrderTree instanceof IfExpr {
+
+  override predicate first(AstNode node) { first(super.getBody(), node) }
+
+  /** Whether this `LoopExpr` captures the `c` completion. */
+  predicate capturesLoopJumpCompletion(LoopJumpCompletion c) {
+    not c.hasLabel()
+    or
+    c.getLabelName() = super.getLabel().getName()
+  }
+
+  override predicate succ(AstNode pred, AstNode succ, Completion c) {
+    // Edge back to the start for final expression and continue expressions
+    last(super.getBody(), pred, c) and
+    (
+      completionIsNormal(c)
+      or
+      c.(LoopJumpCompletion).isContinue() and this.capturesLoopJumpCompletion(c)
+    ) and
+    this.first(succ)
+    or
+    // Edge for exiting the loop with a break expressions
+    last(super.getBody(), pred, c) and
+    c.(LoopJumpCompletion).isBreak() and
+    this.capturesLoopJumpCompletion(c) and
+    succ = this
+  }
+
+  override predicate last(AstNode last, Completion c) {
+    super.last(last, c)
+    or
+    // Any abnormal completions that this loop does not capture should propagate
+    last(super.getBody(), last, c) and
+    not completionIsNormal(c) and
+    not this.capturesLoopJumpCompletion(c)
+  }
+}
+
+class MatchArmTree extends ControlFlowTree instanceof MatchArm {
+  override predicate propagatesAbnormal(AstNode child) { child = super.getExpr() }
+
+  override predicate first(AstNode node) { node = super.getPat() }
+
+  override predicate succ(AstNode pred, AstNode succ, Completion c) {
+    // Edge from pattern to guard/arm if match succeeds.
+    pred = super.getPat() and
+    c.(MatchCompletion).succeeded() and
+    (if super.hasGuard() then first(super.getGuard(), succ) else first(super.getExpr(), succ))
+    or
+    // Edge from guard to arm if the guard succeeds.
+    last(super.getGuard(), pred, c) and
+    first(super.getExpr(), succ) and
+    c.(BooleanCompletion).succeeded()
+  }
+
+  override predicate last(AstNode node, Completion c) {
+    node = super.getPat() and c.(MatchCompletion).failed()
+    or
+    last(super.getGuard(), node, c) and c.(BooleanCompletion).failed()
+    or
+    last(super.getExpr(), node, c)
+  }
+}
+
+class MatchExprTree extends PostOrderTree instanceof MatchExpr {
+  override predicate propagatesAbnormal(AstNode child) { child = super.getABranch().getExpr() }
+
+  override predicate first(AstNode node) { first(super.getExpr(), node) }
+
+  override predicate succ(AstNode pred, AstNode succ, Completion c) {
+    // Edge from the scrutinee to the first arm.
+    last(super.getExpr(), pred, c) and succ = super.getBranch(0).getPat()
+    or
+    // Edge from a failed match/guard in one arm to the beginning of the next arm.
+    exists(int i |
+      last(super.getBranch(i), pred, c) and
+      first(super.getBranch(i + 1), succ) and
+      c.(ConditionalCompletion).failed()
+    )
+    or
+    // Edge from the end of each arm to the match expression.
+    last(super.getBranch(_), pred, c) and succ = this and completionIsSimple(c)
```

Comment:
`completionIsNormal` instead of `completionIsSimple`?

Reply:
This change by itself causes edges to go from the arm patterns and guards to the `match` expression. I propose that we stick with `completionIsSimple` for now and then extend this later such that Boolean completions can propagate from the branch (but _not_ the guard).

Reply:
This also means that we can't allow Boolean completions to be valid for the last node in a `match` arm.

Reply:
That's fine, for now.